src/Control/Effect/Cut.hs

```diff
@@ -102,15 +102,15 @@ instance Monad (CutC m) where
   {-# INLINE (>>=) #-}
 
 instance Fail.MonadFail m => Fail.MonadFail (CutC m) where
-  fail s = CutC (\ _ _ _ -> Fail.fail s)
+  fail s = lift (Fail.fail s)
   {-# INLINE fail #-}
 
 instance MonadFix m => MonadFix (CutC m) where
   mfix f = CutC (\ cons nil _ -> mfix (\ a -> runCutC (f (head a)) (fmap . (:)) (pure []) (pure [])) >>= foldr cons nil)
   {-# INLINE mfix #-}
 
 instance MonadIO m => MonadIO (CutC m) where
-  liftIO io = CutC (\ cons nil _ -> liftIO io >>= flip cons nil)
+  liftIO io = lift (liftIO io)
```

Comment:
I attempted to apply the same approach to `CutC`, but (much to my surprise) the semantics of `cut`/`call` are such that you actually _want_ the left-to-right sequencing of the encoded list formulation. I’m not yet convinced that it _can’t_ be done with a binary tree formulation, but it sure isn’t straightforward.

Reply:
Can you walk me through what’s going on here? I’m not sure I follow the operative difference introduced by this change.

Reply:
The semantics are unchanged, because:

```haskell
lift m = CutC (\ cons nil _ -> m >>= flip cons nil)
```

thus,

```haskell
lift (liftIO io) = CutC (\ cons nil _ -> liftIO io >>= flip cons nil)
```

which is exactly the definition we’ve replaced.